src/rt/minfo.d

```diff
@@ -235,12 +231,82 @@ struct ModuleGroup
             memset(flags, 0, flagbytes);
         }
 
+
+        // build the edges between each module. We may need this for printing,
+        // and also allows avoiding keeping a hash around for module lookups.
+        int[][] edges = (cast(int[]*)malloc((int[]).sizeof * _modules.length))[0 .. _modules.length];
+        {
+            HashTab!(immutable(ModuleInfo)*, int) modIndexes;
+            foreach (i, m; _modules)
+                modIndexes[m] = cast(int) i;
+
+            auto reachable = cast(size_t*) malloc(flagbytes);
+            scope(exit)
+                .free(reachable);
+
+            foreach (i, m; _modules)
+            {
+                // use bit array to prevent duplicates
+                // https://issues.dlang.org/show_bug.cgi?id=16208
+                clearFlags(reachable);
+                // preallocate enough space to store all the indexes
+                int *edge = cast(int*)malloc(int.sizeof * _modules.length);
+                size_t nEdges = 0;
+                foreach (e; m.importedModules)
+                {
+                    if(auto impidx = e in modIndexes)
+                    {
+                        if (*impidx != i)
```

Comment:
You could continue before the hash lookup.
`if (e is m) continue; // self-import`